framework/Models/HybridModel.py

```diff
@@ -0,0 +1,730 @@
+      else:
+        self.raiseAnError(IOError, "Unknown model selection method ", selectionMethod, " is given!")
+
+    if allValid:
+      self.raiseADebug("ROMs  are all valid for given model ", self.modelInstance.name)
+
+    return allValid
+
+  def crowdingDistanceMethod(self, settingDict, varDict):
+    """
+      This function will check the validity of all roms based on the crowding distance method
+      @ In, settingDict, dict, stores the setting information for the crowding distance method
+      @ In, varDict, dict,  is a dictionary that contains the information coming from the sampler,
+        i.e. {'name variable':value}
+      @ Out, allValid, bool, True if the  given sampled point is valid for all roms, otherwise False
+    """
+    allValid = True
+    for romInfo in self.romsDictionary.values():
+      valid = False
+      # generate the data for input parameters
+      paramsList = romInfo['Instance'].getInitParams()['Features']
+      trainInput = np.asarray(self._extractInputs(romInfo['Instance'].trainingSet, paramsList).values())
+      currentInput = np.asarray(self._extractInputs(varDict, paramsList).values())
+      coeffCD = self.computeCDCoefficient(trainInput, currentInput)
+      self.raiseADebug("Crowding Distance Coefficient: ", coeffCD)
+      if coeffCD >= settingDict['threshold']:
+        valid = True
+      romInfo['Valid'] = valid
+      if valid:
+        self.raiseADebug("ROM ",romInfo['Instance'].name, " is valid")
+      else:
+        allValid = False
+
+    return allValid
+
+  def _extractInputs(self,dataIn, paramsList):
+    """
+      Extract the the parameters in the paramsList from the given data object dataIn
+      @ dataIn, Instance or Dict, data object or dictionary contains the input and output parameters
+      @ paramsList, List, List of parameter names
+      @ localInput, dict, dictionary contains the selected input and output parameters and their values
+    """
+    localInput = dict.fromkeys(paramsList, None)
+    if type(dataIn) == dict:
+      for elem, value in dataIn.items():
+        if elem in paramsList:
+          localInput[elem] = np.atleast_1d(value)
+    else:
+      self.raiseAnError(IOError, "The input type '", inputType, "' can not be accepted!")
+
+    return localInput
+
+  def computeCDCoefficient(self, trainSet, newSet):
+    """
+      This function will compute the Crowding distance coefficients among the input parameters
+      @ In, trainSet, numpy.array, array contains values of previous generated input parameters
+      @ In, newSet, numpy.array, array contains values of current generated input parameters
+      @ Out, crowdingDistCoeff, float, the coefficient of crowding distance
+    """
+    totalSet = np.concatenate((trainSet,newSet), axis=1)
+    dim = totalSet.shape[1]
+    distMat = np.zeros((dim, dim))
+    for i in range(dim):
+      for j in range(dim):
+        distMat[i,j] = linalg.norm(totalSet[:,i] - totalSet[:,j])
```

Comment:
this implementation can be quite slow. Any other solution?

Reply:
I think we can pre-compute the norms for existing vectors, and we only need to compute the norms between the new sample and the training sets. In this case, we can reduce loop from dim*dim to dim.